### PR TITLE
Depth breadth

### PR DESCRIPTION
**Pull request by robertwhaskell**

### CODE REVIEW

File: simple_graph.py [hunk 1]
```diff
@@ -27,9 +28,20 @@ def add_edge(self, node1, node2):
             self.graph.setdefault(node1, [node2])
 
     def del_node(self, node):
-        self.graph.pop(node, None)
+        """Delete a node from the graph. Throw a KeyError if the node wasn't
+        in the graph."""
+        try:
+            self.graph.pop(node)
+        except KeyError:
+            raise KeyError("{} does not exist".format(node))
```
Comment (cewing) [inline simple_graph.py:36]: I've commented on a couple of other people's submissions about this.  I think mutating the existing `KeyError` would be better than raising an entirely new one, as the location of the error will now be wrong.

File: simple_graph.py [hunk 2]
```diff
@@ -27,9 +28,20 @@ def add_edge(self, node1, node2):
...
+        for k, v in self.graph.iteritems():
+            if node in v:
+                v.remove(node)
 
     def del_edge(self, node1, node2):
+        """Delete an edge from the graph. Throw a KeyError if the first node
+         wasn't in the graph.  Throw a ValueError if the second node wasn't
+         in the graph."""
```
Comment (cewing) [inline simple_graph.py:44]: Properly formatted docstring would have the first sentence by itself on the first line, and then would put the rest onto later lines.  (see [PEP257](https://www.python.org/dev/peps/pep-0257/))

File: simple_graph.py [hunk 3]
```diff
@@ -37,24 +49,109 @@ def del_edge(self, node1, node2):
...
 
     def adjacent(self, node1, node2):
+        """Return true if an edge points from the first node to the second
+         node. Throw a KeyError if the first node wasn't in the graph."""
         try:
             return node2 in self.graph[node1]
         except KeyError:
             raise KeyError("{} doesn't exist".format(node1))
 
+    def _depth_first_helper(self, node, visited):
```
Comment (markableidinger) [inline simple_graph.py:75]: There's a more efficient method than recursion to do this. Think about a simple data structure that could do the same

File: test_simple_graph.py [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import pytest
 from simple_graph import Graph
+
+
```
Comment (markableidinger) [inline test_simple_graph.py:4]: You should also run appropriate tests for a graph that has an "orphan" node - one that is not connected by an edge to anything else.

File: test_simple_graph.py [hunk 5]
```diff
@@ -11,6 +13,32 @@ def populated():
...
+    g.add_edge(2, 4)
+    g.add_edge(2, 3)
+    g.add_edge(3, 7)
+    g.add_edge(7, 1)
+    g.add_edge(7, 6)
+    return g
+
+
 def test_graph_constructor():
     assert Graph().graph == {}
```
Comment (markableidinger) [inline test_simple_graph.py:44]: Personally, I would prefer that you check is Graph() is an instance of the Graph class, because if you change your implementation that test will always test that your graph is getting made.

File: test_simple_graph.py [hunk 6]
```diff
@@ -41,9 +69,22 @@ def test_add_edge(populated):
...
 def test_del_node(populated):
     populated.del_node(5)
     assert 5 not in populated.graph
+    for k, v in populated.graph.iteritems():
+            assert 5 not in v
+
+
+def test_del_nonexistent_node(populated):
+    with pytest.raises(KeyError):
+        populated.del_node(70)
```
Comment (markableidinger) [inline test_simple_graph.py:87]: Test that your KeyError returns your custom error message - that will ensure that you are hitting YOUR error and not a different KeyError.

This applies to all of your error tests.

File: test_simple_graph.py [hunk 7]
```diff
@@ -75,3 +128,15 @@ def test_adjacent(populated):
 def test_adjacent_none(populated):
     with pytest.raises(KeyError):
         populated.adjacent(58, 28)
+
+
+def test_depth_first(cyclic):
+    nodes = cyclic.depth_first_traversal(1)
```
Comment (markableidinger) [inline test_simple_graph.py:134]: This doesn't test that your traversal is in the correct order, which is kind of the whole point of a traversal. You should work out the correct order on paper and then test it against that.

File: test_simple_graph.py [hunk 8]
```diff
@@ -41,9 +69,22 @@ def test_add_edge(populated):
     assert [7, 5] in populated.edges()
 
 
+def test_add_edge_new_node(populated):
+    populated.add_edge(5, 28)
```
Comment (markableidinger) [inline test_simple_graph.py:73]: I'm a firm believer that every test needs a docstring. Most of your test names are pretty self-evident, so you don't lose a lot by not having docstrings, but this function absolutely requires some added explanation. I had no idea what was going on until I scrolled up to your add_edge function and realized that adding an edge also can add a node.
